src/resolvers/registries/npm-resolver.js

```diff
@@ -187,21 +188,35 @@ export default class NpmResolver extends RegistryResolver {
         const offlineMirrorPath = this.config.getOfflineMirrorPath();
         if (prebuiltVariants[prebuiltName] && offlineMirrorPath) {
           const filename = path.join(offlineMirrorPath, 'prebuilt', prebuiltName + '.tgz');
-          if (shrunk._remote && (await fs.exists(filename))) {
-            shrunk._remote.reference = `file:${filename}`;
-            shrunk._remote.hash = prebuiltVariants[prebuiltName];
+          const {_remote} = shrunk;
+          if (_remote && (await fs.exists(filename))) {
+            _remote.reference = `file:${filename}`;
+            _remote.hash = prebuiltVariants[prebuiltName];
+            _remote.integrity = ssri.fromHex(_remote.hash, 'sha1').toString();
           }
         }
       }
+    }
+    if (shrunk && shrunk._remote && shrunk._remote.integrity) {
+      // if the integrity field does not exist, it needs to be created
       return shrunk;
     }
 
-    const info: ?Manifest = await this.resolveRequest();
+    const desiredVersion = shrunk && shrunk.version ? shrunk.version : null;
+    const info: ?Manifest = await this.resolveRequest(desiredVersion);
```

Comment:
Why do we need to resolve to a different version than before?

Reply:
Before this feature, we would resolve the cached tar file if we had one: https://github.com/yarnpkg/yarn/blob/615ee94b0d88dfbd728d03114aa031cd11e838e1/src/resolvers/registries/npm-resolver.js#L196
Here, we add the condition that said cached file has to have an integrity field: https://github.com/yarnpkg/yarn/pull/5042/files#diff-3aebe0cb09fdaf76a6f7303534a3ee26R200
If it doesn't, we want to fetch the package again from the npm repository in order to get that field from there to update the lockfile for future checksum checks. When we do this, we want to make sure we get the exact same version as specified in the lockfile and not (as the method would without adding the `desiredVersion` argument) a version that satisfies the semver range

Reply:
I'll start with the short and honest answer: because it would require a great deal more changes than this already monstrous and hard-to-swallow PR. :)
Right now, what this PR does is create compatibility with the npm integrity feature. It mirrors the integrity from the npm repository (and creates it when publishing to npm through yarn). All of the integrity checks are performed against fields that at some point were in a (npm) registry json file. This whole feature is built to support that. For everything else, we still have the `hash` field which - while not perfect - can keep serving more than just backwards compatibility for now. This feature does give a good infrastructure to eventually replace the `hash` field (maybe in a new lockfile version?) though. When it does that, it can replace all the other roles of the `hash` field such as being a checksum for non-registry packages.
I hope we see eye to eye on this one?

Reply:
Makes sense 👍